core/grafana-data-source/src/server.rs

```diff
@@ -0,0 +1,134 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use serde::{Serialize, de::DeserializeOwned};
+use hyper::{Body, Request, Response, header, service::service_fn, Server};
+use futures::{future, Future, stream::Stream};
+use chrono::{Duration, Utc};
+use stream_cancel::StreamExt;
+use crate::{METRICS, util, types::{Target, Query, TimeseriesData}};
+
+type GenericError = Box<dyn std::error::Error + Send + Sync>;
+type ResponseFuture = Box<dyn Future<Item=Response<Body>, Error=GenericError> + Send>;
+
+fn api_response(req: Request<Body>) -> ResponseFuture {
+	match req.uri().path() {
+		"/" => Box::new(future::ok(Response::new(Body::empty()))),
+		"/search" => map_request_to_response(req, |target: Target| {
+			// Filter and return metrics relating to the target
+			METRICS.read()
+				.keys()
+				.filter(|key| key.starts_with(&target.target))
+				.cloned()
+				.collect::<Vec<_>>()
+		}),
+		"/query" => {
+			map_request_to_response(req, |query: Query| {
+				let metrics = METRICS.read();
+
+				// Return timeseries data related to the specified metrics
+				query.targets.iter()
+					.map(|target| {
+						let datapoints = metrics.get(target.target.as_str())
+							.map(|metric| {
+								let from = util::find_index(&metric, query.range.from);
+								let to = util::find_index(&metric, query.range.to);
+
+								let metric = &metric[from .. to];
+
+								if metric.len() > query.max_datapoints {
+									// Avoid returning more than `max_datapoints` (mostly to stop
+									// the web browser from having to do a ton of work)
+									util::select_points(metric, query.max_datapoints)
+								} else {
+									metric.to_vec()
+								}
+							})
+							.unwrap_or_else(Vec::new);
+
+						TimeseriesData {
+							target: target.target.clone(), datapoints
+						}
+					})
+					.collect::<Vec<_>>()
+			})
+		},
+		_ => Box::new(future::ok(Response::new(Body::empty()))),
+	}
+}
+
+fn map_request_to_response<Req, Res, T>(req: Request<Body>, transformation: T) -> ResponseFuture
+	where
+		Req: DeserializeOwned,
+		Res: Serialize,
+		T: Fn(Req) -> Res + Send + Sync + 'static
+{
+	Box::new(req.into_body()
+		.concat2()
+		.from_err()
+		.and_then(move |entire_body| {
+			let req = serde_json::from_slice(entire_body.as_ref())?;
+			let res = transformation(req);
+
+			let string = serde_json::to_string(&res)?;
+
+			Response::builder()
+				.header(header::CONTENT_TYPE, "application/json")
+				.body(Body::from(string))
+				.map_err(|e| e.into())
+		})
+	)
+}
+
+/// Start the data source server.
+///
+/// The server shuts down cleanly when `shutdown` resolves.
+pub fn run_server<F>(address: &std::net::SocketAddr, shutdown: F) -> impl Future<Item=(), Error=()>
```

Comment:
I'm not sure how hyper works exactly (as it's quite un-idiomatic in this regard), I don't think we should pass a `shutdown` parameter here. Instead, we should do `run_server(...).select(exit)` in the service. When `exit` triggers, the `Future` here would then be destroyed and would shut down the server.

Reply:
The service has some APIs for avoiding explicit `select`ion because it is error-prone to omission. I think the API is called `spawn_task`, where we pass a future that we want to be combined with the overall `exit`.

Reply:
I'm happy to use something other than hyper if it would help, I just chose it because it's already used in substrate.

Reply:
Where should `spawn_task` be called in this case? I see it only being used in the node currently.

Reply:
What I was suggesting is that instead of doing [this](https://github.com/paritytech/substrate/pull/3913/files#diff-1c0bedf4d6831513aa1122c56b1003bdR1118), you do something like this:

```rust
grafana_data_source::run_server(&port).select(exit.clone())
```

And the `exit` parameter would no longer be passed to `run_server`.

But as I mentioned I would have to look into the docs of hyper to see if it properly shuts down when you kill the future, so you tell me if that makes sense.

Reply:
Yep, `ctrl`-`c` seems to stall forever if I use `select` instead. Plus, the code ends up looking more like `.select(exit.clone()).map(|_| ()).map_err(|_| ())`, which is less pleasant.